jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/generator/HeadersGenerator.java

```diff
@@ -63,82 +60,65 @@ public int generateHeaders(ByteBufferPool.Accumulator accumulator, int streamId,
         if (priority != null)
             flags = Flags.PRIORITY;
 
+        // TODO Look for a way of not allocating a large buffer here.
+        //      Possibly the hpack encoder could be changed to take the accumulator, but that is a lot of changes.
+        //      Alternately, we could ensure the accumulator has maxFrameSize space
+        //      So long as the buffer is not sliced into continuations, it at least should be available to aggregate
+        //      subsequent frames into... but likely only a frame header followed by an accumulated data frame.
+        //      It might also be good to be able to split the table into continuation frames as it is generated?
         RetainableByteBuffer hpack = encode(encoder, metaData, getMaxFrameSize());
-        ByteBuffer hpackByteBuffer = hpack.getByteBuffer();
-        int hpackLength = hpackByteBuffer.position();
-        BufferUtil.flipToFlush(hpackByteBuffer, 0);
+        BufferUtil.flipToFlush(hpack.getByteBuffer(), 0);
+        int hpackLength = hpack.remaining();
 
         // Split into CONTINUATION frames if necessary.
         if (maxHeaderBlockFragment > 0 && hpackLength > maxHeaderBlockFragment)
         {
+            int start = accumulator.remaining();
             if (endStream)
                 flags |= Flags.END_STREAM;
 
-            int length = maxHeaderBlockFragment;
-            if (priority != null)
-                length += PriorityFrame.PRIORITY_LENGTH;
-
-            RetainableByteBuffer header = generateHeader(FrameType.HEADERS, length, flags, streamId);
-            ByteBuffer headerByteBuffer = header.getByteBuffer();
-            generatePriority(headerByteBuffer, priority);
-            BufferUtil.flipToFlush(headerByteBuffer, 0);
-            accumulator.append(header);
-            hpackByteBuffer.limit(maxHeaderBlockFragment);
-            accumulator.append(RetainableByteBuffer.wrap(hpackByteBuffer.slice()));
+            int length = maxHeaderBlockFragment + (priority == null ? 0 : PriorityFrame.PRIORITY_LENGTH);
 
-            int totalLength = Frame.HEADER_LENGTH + length;
+            // generate first fragment with as HEADERS with possible priority
+            generateHeader(accumulator, FrameType.HEADERS, length, flags, streamId);
+            generatePriority(accumulator, priority);
+            accumulator.add(hpack.slice(maxHeaderBlockFragment));
```

Comment:
If `hpack.slice()` retains (like the RBB impl does) then the RBB returned by `slice()` should be released here. If it does not retains (like the RBB.DynamicCapacity impl does) then this code is fine... but it's breaking the `Retainable` contract.

Reply:
The slice is retained, but then added  to the accumulator, so the contract is that the accumulator take over responsibility for releasing the slice.    If I had appended, then I would need to release here.

Reply:
@gregw that is the opposite of the documented behavior that `Retainable`s must have.

Reply:
@sbordet you explicitly asked me to provide the add semantic so we can "give" a RBB and not have to release it ourselves.  The documented behaviour for Retainable is described as "typical" and as a "general rule", should I think is equivalent to a SHOULD rather than a MUST.    The divergence from the norm is well documented in the signature of the `add` method.

We could instead do:
```suggestion
            RetainableByteBuffer hpackSlice = hpack.slice(maxHeaderBlockFragment);
            accumulator.append(hpackSlice);
            hpackSlice.release();
```
But that is more verbose and means the reference counter is needlessly incremented and decremented, then we end up at the same place.